### PR TITLE
Ensure correct rate limiting for bad clients.

### PR DESCRIPTION
**Pull request by felixdesouza**
**Goals (and why)**:
I previously incorrectly rate limited, this now takes into account shards and _local_ threads.

**Implementation Description (bullets)**:
rate limit is now:

    min(threads, shards) * 2_PERMITS_PER_SECOND / shards

Rationale being:
* with 1 thread, n shards, you will visit a shard every `n * pausetime` seconds.
* with 2 threads, n shards, you will visit a shard every `n * pausetime / 2` seconds.
* ...
* with m threads, n shards, you will visit a shard every `n * pausetime / m` seconds.

The reciprocal of the above is the permits per second.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a "circular" test which closer resembles targeted sweep.

**Priority (whenever / two weeks / yesterday)**:
ASAP, want to release quickly

### CODE REVIEW

File: changelog/@unreleased/pr-4110.v2.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+type: improvement
+improvement:
+  description: Correctly rate limit bad targeted sweep clients.
```
Comment (jeremyk-91) [inline changelog/@unreleased/pr-4110.v2.yml]: I think we should say what `Correctly` means, it's a bit unclear exactly what number we're aiming for. I gather that for each strategy, we want to allow 2 requests per second for the smaller of `(shards, threads)` (to kind-of simulate the old behaviour)? Though not sure if `threads` is intended to be the sum across all nodes (which is a closer simulation of old behaviour) or the number on each node (more conservative)

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/TargetedSweepLockDecorator.java [hunk 2]
```diff
@@ -105,14 +133,24 @@ public OrderedLocks decorate(List<LockDescriptor> descriptors, List<AsyncLock> l
...
     }
 
+    private AsyncLock targetedSweepLock(LockDescriptor descriptor, AsyncLock lock) {
+        ByteString descriptorByteString = ByteString.copyFrom(descriptor.getBytes());
+        if (descriptorByteString.endsWith(CONSERVATIVE_SUFFIX)) {
+            return targetedSweepLockByDelegate.get(Maps.immutableEntry(SweepStrategy.CONSERVATIVE, lock));
+        } else if (descriptorByteString.endsWith(THOROUGH_SUFFIX)) {
+            return targetedSweepLockByDelegate.get(Maps.immutableEntry(SweepStrategy.THOROUGH, lock));
+        } else {
+            return lock;
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/TargetedSweepLockDecorator.java]: based on what we talked about last time, you'd have to be doing janky stuff to get this going

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        @Value.Default
+        @Value.Parameter
+        default ShardAndThreadConfig config() {
+            return ShardAndThreadConfig.defaultConfig();
+        }
+    }
+
+    @JsonDeserialize(as = ImmutableShardAndThreadConfig.class)
+    @JsonSerialize(as = ImmutableShardAndThreadConfig.class)
+    @JsonIgnoreProperties(ignoreUnknown = true)
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java]: nope, from my template

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+        default double permitsPerSecond(int threads, int shards) {
+            int concurrency = Math.min(threads, shards);
+            return ((double) LOCK_ACQUIRES_PER_SECOND) / shards * concurrency;
+        }
+
+        static ShardAndThreadConfig defaultConfig() {
+            return ImmutableShardAndThreadConfig.builder().build();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java]: kebab-case it is

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        @Value.Default
+        default int thoroughThreads() {
+            return 1;
+        }
+
+        @Value.Default
+        default int thoroughShards() {
+            return 1;
+        }
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java]: oop my bad!

[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/config/TargetedSweepLockControlConfig.java]: I thought there was an analogue where we had conservativeThreads + thoroughThreads

File: timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceTargetedSweepRateLimitingTest.java [hunk 6]
```diff
@@ -117,21 +120,82 @@ public void full_ete() throws InterruptedException {
...
+        Thread.sleep(15_000);
+        combinedFuture.cancel(true);
+
+        assertThat(meanRate(nonRateLimited))
+                .as("non rate limited client isn't rate limited when we go in a circle like targeted sweep")
+                .satisfies(AsyncTimelockServiceTargetedSweepRateLimitingTest::isNotRateLimited);
+
+        assertThat(ImmutableList.of(meanRate(rateLimitedThorough), meanRate(rateLimitedThorough2)))
+                .as("thorough is rate limited with 8 shards and 2 threads, each 'thread' should get 2 per second per "
+                        + "thread")
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceTargetedSweepRateLimitingTest.java]: hmm, odd

**Pull request merged**